File: kernel/bootstrap/logger.rb
module Rubinius
  class Logger
    attr_accessor :name
    attr_accessor :format

    def self.allocate
      Rubinius.primitive :logger_allocate
      raise PrimitiveFailure, "Rubinius::Logger.allocate failed"
    end

    def self.system
      @system_logger ||= new "system"
    end

    def self.log_exception(message, e)
      log = system

      log.error message
      STDERR.puts message, "\n"

      msg = "#{e.message} (#{e.class})"
      log.error msg
      STDERR.puts msg, "\nBacktrace:\n\n"

      color = STDERR.tty?
      exception = e

      while exception
        if custom_trace = exception.custom_backtrace
          custom_trace.each do |line|
            log.error line
            STDERR.puts line
          end
        end

        exception.backtrace.each { |line| log.error line }
        STDERR.puts exception.awesome_backtrace.show("\n", color)

        if exception = exception.parent
          msg = "Caused by: #{exception.message} (#{exception.class})"
          log.error msg
          STDERR.puts "\n", msg, "\n"
        end
      end
    end

    def initialize(name)
      @name = name
      @format = "#{name}: %s"
    end

    def write(message)
      Rubinius.primitive :logger_write
      raise PrimitiveFailure, "Rubinius::Logger#write primitive failed"
    end

    def fatal(message)
      Rubinius.primitive :logger_fatal
      raise PrimitiveFailure, "Rubinius::Logger#fatal primitive failed"
    end

    def error(message)
      Rubinius.primitive :logger_error
      raise PrimitiveFailure, "Rubinius::Logger#error primitive failed"
    end

    def warn(message)
      Rubinius.primitive :logger_warn
      raise PrimitiveFailure, "Rubinius::Logger#warn primitive failed"
    end

    def info(message)
      Rubinius.primitive :logger_info
      raise PrimitiveFailure, "Rubinius::Logger#info primitive failed"
    end

    def debug(message)
      Rubinius.primitive :logger_debug
      raise PrimitiveFailure, "Rubinius::Logger#debug primitive failed"
    end
  end
end

File: vm/builtin/logger.cpp
#include "builtin/class.hpp"
#include "builtin/logger.hpp"
#include "builtin/string.hpp"

#include "util/logger.hpp"

#include "ontology.hpp"

namespace rubinius {
  using namespace utilities;

  void Logger::init(STATE) {
    GO(logger).set(ontology::new_class_under(state, "Logger", G(rubinius)));
    G(logger)->set_object_type(state, LoggerType);
  }

  Logger* Logger::create(STATE) {
    Logger* logger = state->new_object<Logger>(G(logger));
    return logger;
  }

  Logger* Logger::allocate(STATE, Object* self) {
    Logger* logger = create(state);
    logger->klass(state, as<Class>(self));
    return logger;
  }

  Object* Logger::write(STATE, String* message) {
    logger::write(format()->c_str(state), message->c_str(state));
    return cNil;
  }

  Object* Logger::fatal(STATE, String* message) {
    logger::fatal(format()->c_str(state), message->c_str(state));
    return cNil;
  }

  Object* Logger::error(STATE, String* message) {
    logger::error(format()->c_str(state), message->c_str(state));
    return cNil;
  }

  Object* Logger::warn(STATE, String* message) {
    logger::warn(format()->c_str(state), message->c_str(state));
    return cNil;
  }

  Object* Logger::info(STATE, String* message) {
    logger::info(format()->c_str(state), message->c_str(state));
    return cNil;
  }

  Object* Logger::debug(STATE, String* message) {
    logger::debug(format()->c_str(state), message->c_str(state));
    return cNil;
  }
}
